### PR TITLE
fix a bug when using fp16 training & gradient clipping

### PR DESCRIPTION
**Pull request by YKX-A**
Fixes # .

Changes proposed in this pull request:

- A bug that incorrectly clips scaled gradients when we use mixed-precision training and gradient clipping at the same time. So I modify the logic of gradient clipping. (see also #5413)
(@epwalsh, Come here to see if it's ok.)
## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.
- [ ] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

**Comment by epwalsh** [comment]
@YKX-A can you run `black .` from the repo root to fix formatting.

File: allennlp/training/gradient_descent_trainer.py [hunk 1]
```diff
@@ -349,6 +350,22 @@ def _pytorch_model(self):
...
+    def clip_gradient(self):
+        """
+        Performs gradient clipping.
+        If the model is in mixed precision training, we would first unscale the gradient.
+        """
+        if self._grad_clipping is not None:
+            # 1. We have to unscale the gradient before clipping
+            if self._scaler is not None:
+                optimizer_state = self._scaler._per_optimizer_states[id(self.optimizer)]
+                # 2. The `unscale_` should't be performed more than once per optimizer per step call,
```
Comment (epwalsh) [inline allennlp/training/gradient_descent_trainer.py:362]:
```suggestion
                # 2. The `unscale_` shouldn't be performed more than once per optimizer per step call,
```

**Review by epwalsh — COMMENTED** [review]
This looks great other than the minor formatting issues!

File: allennlp/training/gradient_descent_trainer.py (hunk 1 above)
Comment (YKX-A) [reply allennlp/training/gradient_descent_trainer.py]: Oh, that's my ignorance 😰. Is it correct now?

**Comment by YKX-A** [comment]
@epwalsh
